Default value of build_base leads to races when running tests concurrently

In [miri](https://github.com/solson/miri), we use compiletest in multiple `#[test]` functions, so they are executed in parallel. We see a lot of test failures like
```
error: compilation failed!

status: exit code: 101

command: target/release/miri tests/run-pass/aux_test.rs -L /tmp --target=x86_64-unknown-linux-gnu -Z dump-mir=all -Z mir-opt-level=3 -Z dump-mir-dir=/tmp/aux_test -L /tmp/aux_test.stage-id.mir-opt.libaux -C prefer-dynamic -o /tmp/aux_test.stage-id -Zmir-opt-level=0 -Zmir-emit-validate=1

stdout:

------------------------------------------

------------------------------------------

stderr:

------------------------------------------

error: linking with `cc` failed: exit code: 1

  |

  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/aux_test.0.o" "-o" "/tmp/aux_test.stage-id" "/tmp/aux_test.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-L" "/tmp" "-L" "/tmp/aux_test.stage-id.mir-opt.libaux" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/tmp/aux_test.stage-id.mir-opt.libaux" "-l" "dep" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-l" "std-ab8512bb715ab82f" "-Wl,-Bstatic" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-53095306c4bf2e4a.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"

  = note: /usr/bin/ld: cannot find /tmp/aux_test.0.o: No such file or directory

          /usr/bin/ld: cannot find /tmp/aux_test.crate.allocator.o: No such file or directory

          collect2: ld returned 1 exit status
```

From the behavior of the bug (it only occurs spuriously), this feels a lot like a race condition.
(It also shows that compiletest messes with `/tmp` really badly, rather than creating its own subdirectory, which I would have expected.)

We track this in miri as https://github.com/solson/miri/issues/305

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -17,3 +17,7 @@ name = "compiletest_rs"
 [dependencies]
 log = "0.3.5"
 rustc-serialize = "0.3"
+tempdir = { version = "0.3", optional = true }
+
+[features]
+tmp = ["tempdir"]

--- a/src/common.rs
+++ b/src/common.rs
@@ -232,9 +232,25 @@ impl Config {
 #[cfg(feature = "tmp")]
 mod config_tempdir {
     use tempdir;
+    use std::ops;
+
     pub struct ConfigWithTemp {
         pub config: super::Config,
         pub tempdir: tempdir::TempDir,
+    }
+
+    impl ops::Deref for ConfigWithTemp {
+        type Target = super::Config;
+
+        fn deref(&self) -> &Self::Target {
+            &self.config
+        }
+    }
+
+    impl ops::DerefMut for ConfigWithTemp {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.config
+        }
     }
 }
 

--- a/src/common.rs
+++ b/src/common.rs
@@ -224,8 +224,7 @@ impl Config {
 fn tempdir() -> PathBuf {
     tempdir::TempDir::new("compiletest")
         .expect("failed to create temporary directory")
-        .path()
-        .to_path_buf()
+        .into_path()
 }
 
 #[cfg(not(feature = "tmp"))]

--- a/src/common.rs
+++ b/src/common.rs
@@ -15,9 +15,6 @@ use std::str::FromStr;
 use std::path::PathBuf;
 use rustc;
 
-#[cfg(feature = "tmp")]
-use tempdir;
-
 #[derive(Clone, Copy, PartialEq, Debug)]
 pub enum Mode {
     CompileFail,
@@ -218,19 +215,29 @@ impl Config {
 
         self.target_rustcflags = Some(flags);
     }
+
+    #[cfg(feature = "tmp")]
+    pub fn tempdir(mut self) -> config_tempdir::ConfigWithTemp {
+        use tempdir;
+        let tmp = tempdir::TempDir::new("compiletest")
+            .expect("failed to create temporary directory");
+        self.build_base = tmp.path().to_owned();
+        config_tempdir::ConfigWithTemp {
+            config: self,
+            tempdir: tmp,
+        }
+    }
 }
 
 #[cfg(feature = "tmp")]
-fn tempdir() -> PathBuf {
-    tempdir::TempDir::new("compiletest")
-        .expect("failed to create temporary directory")
-        .into_path()
+mod config_tempdir {
+    use tempdir;
+    pub struct ConfigWithTemp {
+        pub config: super::Config,
+        pub tempdir: tempdir::TempDir,
+    }
 }
 
-#[cfg(not(feature = "tmp"))]
-fn tempdir() -> PathBuf {
-    env::temp_dir()
-}
 
 impl Default for Config {
     fn default() -> Config {
@@ -247,7 +254,7 @@ impl Default for Config {
             force_valgrind: false,
             llvm_filecheck: None,
             src_base: PathBuf::from("tests/run-pass"),
-            build_base: tempdir(),
+            build_base: env::temp_dir(),
             stage_id: "stage-id".to_owned(),
             mode: Mode::RunPass,
             run_ignored: false,

--- a/src/common.rs
+++ b/src/common.rs
@@ -15,6 +15,9 @@ use std::str::FromStr;
 use std::path::PathBuf;
 use rustc;
 
+#[cfg(feature = "tmp")]
+use tempdir;
+
 #[derive(Clone, Copy, PartialEq, Debug)]
 pub enum Mode {
     CompileFail,
@@ -217,6 +220,19 @@ impl Config {
     }
 }
 
+#[cfg(feature = "tmp")]
+fn tempdir() -> PathBuf {
+    tempdir::TempDir::new("compiletest")
+        .expect("failed to create temporary directory")
+        .path()
+        .to_path_buf()
+}
+
+#[cfg(not(feature = "tmp"))]
+fn tempdir() -> PathBuf {
+    env::temp_dir()
+}
+
 impl Default for Config {
     fn default() -> Config {
         let platform = rustc::session::config::host_triple().to_string();
@@ -232,7 +248,7 @@ impl Default for Config {
             force_valgrind: false,
             llvm_filecheck: None,
             src_base: PathBuf::from("tests/run-pass"),
-            build_base: env::temp_dir(),
+            build_base: tempdir(),
             stage_id: "stage-id".to_owned(),
             mode: Mode::RunPass,
             run_ignored: false,

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -19,6 +19,8 @@ extern crate test;
 extern crate rustc;
 extern crate rustc_serialize;
 
+#[cfg(feature = "tmp")] extern crate tempdir;
+
 #[macro_use]
 extern crate log;
 

--- a/test-project/Cargo.toml
+++ b/test-project/Cargo.toml
@@ -7,3 +7,4 @@ authors = ["Thomas Jespersen <laumann.thomas@gmail.com>"]
 
 [dev-dependencies.compiletest_rs]
 path = ".."
+features = ["tmp"]

--- a/test-project/tests/tests.rs
+++ b/test-project/tests/tests.rs
@@ -10,8 +10,9 @@ fn run_mode(mode: &'static str) {
     config.mode = cfg_mode;
     config.src_base = PathBuf::from(format!("tests/{}", mode));
     config.link_deps();
+    let wrapped_config = config.tempdir();
 
-    compiletest::run_tests(&config);
+    compiletest::run_tests(&wrapped_config.config);
 }
 
 #[test]

--- a/test-project/tests/tests.rs
+++ b/test-project/tests/tests.rs
@@ -4,15 +4,14 @@ use std::path::PathBuf;
 
 fn run_mode(mode: &'static str) {
 
-    let mut config = compiletest::Config::default();
+    let mut config = compiletest::Config::default().tempdir();
     let cfg_mode = mode.parse().expect("Invalid mode");
 
     config.mode = cfg_mode;
     config.src_base = PathBuf::from(format!("tests/{}", mode));
     config.link_deps();
-    let wrapped_config = config.tempdir();
 
-    compiletest::run_tests(&wrapped_config.config);
+    compiletest::run_tests(&config);
 }
 
 #[test]

